### PR TITLE
Add PadFusion transformation

### PR DESCRIPTION
**Pull request by mateusztabaka**
Ticket: 46482

### CODE REVIEW

File: inference-engine/src/transformations/src/transformations/common_optimizations/pad_fusion.cpp [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+                         const std::shared_ptr<opset5::Constant>& pads_begin, const std::shared_ptr<opset5::Constant>& pads_end) {
+    if (!pad)
+        return false;
+    if (!node)
+        return false;
+    if (pad->get_pad_mode() != op::PadMode::CONSTANT)
+        return false;
+
+    if (!pad_value_const)
+        return false;
```
Comment (ilyachur) [inline inference-engine/src/transformations/src/transformations/common_optimizations/pad_fusion.cpp:31]: Can we fuse `if`s?

Comment (mateusztabaka) [reply inference-engine/src/transformations/src/transformations/common_optimizations/pad_fusion.cpp:31]: I can fuse lines 23 & 27. I think fusing other will be less readable.

File: inference-engine/src/transformations/src/transformations/common_optimizations/pad_fusion.cpp [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+}
+
+NGRAPH_RTTI_DEFINITION(pass::PadFusionAvgPool, "PadFusionAvgPool", 0);
+
+pass::PadFusionAvgPool::PadFusionAvgPool() {
+    auto data_pattern = pattern::any_input();
+    auto pads_begin_pattern = pattern::wrap_type<opset5::Constant>();
+    auto pads_end_pattern = pattern::wrap_type<opset5::Constant>();
+    auto pad_value_pattern = pattern::wrap_type<opset5::Constant>();
+    auto pad_node_pattern = pattern::wrap_type<opset5::Pad>({data_pattern, pads_begin_pattern, pads_end_pattern, pad_value_pattern});
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/pad_fusion.cpp:72]: Check hat pad has exactly one consumer
```suggestion
    auto pad_node_pattern = pattern::wrap_type<opset5::Pad>({data_pattern, pads_begin_pattern, pads_end_pattern, pad_value_pattern}, pattern::consumers_count(1));
```

File: inference-engine/src/transformations/src/transformations/common_optimizations/pad_fusion.cpp [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+        Shape new_pads_begin, new_pads_end;
+        std::tie(new_pads_begin, new_pads_end) = new_pad_values(pads_begin, pads_end, avg_pool);
+        auto new_avg_pool = std::make_shared<opset5::AvgPool>(data, avg_pool->get_strides(),
+                                                              new_pads_begin, new_pads_end,
+                                                              avg_pool->get_kernel(), false,
+                                                              avg_pool->get_rounding_type(),
+                                                              op::PadType::EXPLICIT);
+        new_avg_pool->set_friendly_name(avg_pool->get_friendly_name());
+
+        copy_runtime_info(avg_pool, new_avg_pool);
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/pad_fusion.cpp:95]: You need to copy rt info from pad to new avg_pool also.
```suggestion
        copy_runtime_info({pad, avg_pool}, new_avg_pool);
```

File: inference-engine/src/transformations/src/transformations/common_optimizations/pad_fusion.cpp [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+
+
+pass::PadFusionConvolution::PadFusionConvolution() {
+    auto data_pattern = pattern::any_input();
+    auto filter_pattern = pattern::any_input();
+    auto pads_begin_pattern = pattern::wrap_type<opset5::Constant>();
+    auto pads_end_pattern = pattern::wrap_type<opset5::Constant>();
+    auto pad_value_pattern = pattern::wrap_type<opset5::Constant>();
+    auto pad_node_pattern = pattern::wrap_type<opset5::Pad>({data_pattern, pads_begin_pattern, pads_end_pattern, pad_value_pattern});
+    auto conv_pattern = pattern::wrap_type<opset5::Convolution>({pad_node_pattern, filter_pattern});
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/pad_fusion.cpp:173]: I think we need to covert all convolution types then:
* Convolution
* GroupConvolution
* ConvolutionBackPropData
* GroupConvolutionBackPropData

**Review by GlebKazantaev — CHANGES_REQUESTED** [review]

**Review by mbencer — APPROVED** [review]

**Review by postrational — APPROVED** [review]

**Review by ilyachur — APPROVED** [review]

**Review by tomdol — APPROVED** [review]

File: inference-engine/src/transformations/src/transformations/common_optimizations/pad_fusion.cpp [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+                   pads_end_val.begin() + 2, node_pads_end.begin(),
+                   [] (size_t a, size_t b) -> size_t { return a + b; });
+
+    return std::make_tuple(node_pads_begin, node_pads_end);
+}
+
+NGRAPH_RTTI_DEFINITION(pass::PadFusionAvgPool, "PadFusionAvgPool", 0);
+
+pass::PadFusionAvgPool::PadFusionAvgPool() {
+    auto data_pattern = pattern::any_input();
```
Comment (ilyachur) [inline inference-engine/src/transformations/src/transformations/common_optimizations/pad_fusion.cpp:75]: Please add CC macros. You can find examples in the PR https://github.com/openvinotoolkit/openvino/pull/3795

**Review by GlebKazantaev — CHANGES_REQUESTED** [review]

**Comment by ilyachur** [comment]
Please resolve merge conflicts

**Comment by ilyachur** [comment]
@GlebKazantaev Do you have any comments?

**Review by GlebKazantaev — APPROVED** [review]

**Pull request merged**